lib/utils/platform/iot_clock_afr.c

```diff
@@ -0,0 +1,215 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file iot_clock_afr.c
+ * @brief Implementation of the functions in iot_clock.h for Amazon FreeRTOS systems.
+ */
+
+/* Build using a config header, if provided. */
+#ifdef IOT_CONFIG_FILE
+    #include IOT_CONFIG_FILE
+#endif
+
+/* Platform clock include. */
+#include "platform/iot_clock.h"
+#include "task.h"
+
+/* Configure logs for the functions in this file. */
+#ifdef IOT_LOG_LEVEL_PLATFORM
+    #define _LIBRARY_LOG_LEVEL        IOT_LOG_LEVEL_PLATFORM
+#else
+    #ifdef IOT_LOG_LEVEL_GLOBAL
+        #define _LIBRARY_LOG_LEVEL    IOT_LOG_LEVEL_GLOBAL
+    #else
+        #define _LIBRARY_LOG_LEVEL    IOT_LOG_NONE
+    #endif
+#endif
+
+#define _LIBRARY_LOG_NAME    ( "CLOCK" )
+#include "iot_logging_setup.h"
+
+/*-----------------------------------------------------------*/
+/*
+ * Time conversion constants.
+ */
+#define _NANOSECONDS_PER_SECOND         ( 1000000000 ) /**< @brief Nanoseconds per second. */
+#define _NANOSECONDS_PER_MILLISECOND    ( 1000000 )    /**< @brief Nanoseconds per millisecond. */
+#define _MILLISECONDS_PER_SECOND        ( 1000 )       /**< @brief Milliseconds per second. */
+#define _MILLISECONDS_PER_TICK          ( _MILLISECONDS_PER_SECOND / configTICK_RATE_HZ ) /**< Milliseconds per FreeRTOS tick. */
+
+/*-----------------------------------------------------------*/
+
+
+/*-----------------------------------------------------------*/
+
+/*  Private Callback function for timer expiry, delegate work to a Task to free 
+    up the timer task for managing other timers */
+static void prvTimerCallback( TimerHandle_t xTimerHandle )
+{
+    _timerInfo_t * pxTimer = ( _timerInfo_t * ) pvTimerGetTimerID( xTimerHandle );
+
+    /* The value of the timer ID, set in timer_create, should not be NULL. */
+    configASSERT( pxTimer != NULL );
+
+    /* Restart the timer if it is periodic. */
+    if( pxTimer->xTimerPeriod > 0 )
+    {
+        xTimerChangePeriod( xTimerHandle, pxTimer->xTimerPeriod, 0 );
+    }
+
+    /* Call timer Callback from this task */
+    pxTimer->threadRoutine( ( void * ) pxTimer->pArgument);
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotClock_GetTimestring( char * const pBuffer,
+                                size_t bufferSize,
+                                size_t * const pTimestringLength )
+{
+    uint64_t milliSeconds = IotClock_GetTimeMs( );
```

Comment:
Check parameters?

Reply:
Yes! I kept the checks and behavior identical to the posix implementation. We should add checks to both, probably assert on both pointers is the best we can do here? Agree?

Also this function prints time as Ms since startup, while the posix one uses strftime, to here is a very different behavior, we discussed this before, but I wanted to point it out in review.